File: migrate_roles.js
const Database = require('better-sqlite3');
const db = new Database('pocketbase/pb_data/data.db');

console.log('Starting migration...');

const usersCollection = db.prepare("SELECT * FROM _collections WHERE name = 'users'").get();
if (usersCollection) {
    let schema = JSON.parse(usersCollection.schema);
    const hasRole = schema.some(f => f.name === 'role');

    if (!hasRole) {
        schema.push({
            system: false,
            id: 'users_role_field',
            name: 'role',
            type: 'text',
            required: false,
            presentable: false,
            unique: false,
            options: { min: null, max: null, pattern: '' }
        });

        db.prepare('UPDATE _collections SET schema = ? WHERE id = ?').run(JSON.stringify(schema), usersCollection.id);
        console.log('Added role field to users collection schema.');
    } else {
        console.log('Role field already exists in schema.');
    }
}

// Update existing users in the actual users table
try {
    const info = db.prepare("UPDATE users SET role = 'user' WHERE role IS NULL OR role = ''").run();
    console.log(`Updated ${info.changes} users to have 'user' role.`);
} catch (e) {
    console.log('Could not set default roles, column may not be created yet by PocketBase.');
}

// Now let's print the collections we need to update API rules for
const collections = db.prepare("SELECT name, listRule, viewRule, createRule, updateRule, deleteRule FROM _collections WHERE name IN ('users', 'links', 'clicks')").all();
console.log('\nCurrent API Rules:');
console.log(JSON.stringify(collections, null, 2));

console.log('\nMigration complete.');

File: migrate_roles.py
import sqlite3
import json

db_path = 'pocketbase/pb_data/data.db'

print('Starting migration...')
conn = sqlite3.connect(db_path)
cursor = conn.cursor()

# Get users collection schema
cursor.execute("SELECT id, schema FROM _collections WHERE name = 'users'")
row = cursor.fetchone()

if row:
    collection_id = row[0]
    schema = json.loads(row[1])
    
    has_role = any(f.get('name') == 'role' for f in schema)
    
    if not has_role:
        schema.append({
            'system': False,
            'id': 'users_role_field',
            'name': 'role',
            'type': 'text',
            'required': False,
            'presentable': False,
            'unique': False,
            'options': { 'min': None, 'max': None, 'pattern': '' }
        })
        
        cursor.execute("UPDATE _collections SET schema = ? WHERE id = ?", (json.dumps(schema), collection_id))
        conn.commit()
        print('Added role field to users collection schema.')
    else:
        print('Role field already exists in schema.')

# Try to update users in the table (might fail if we haven't restarted PB to sync schema)
try:
    cursor.execute("UPDATE users SET role = 'user' WHERE role IS NULL OR role = ''")
    conn.commit()
    print(f"Updated {cursor.rowcount} users to have 'user' role.")
except Exception as e:
    print(f"Notice: Could not set default roles (table might need PB restart to create column): {e}")

# Also let's find the main admin user and make them an admin if they exist
try:
    cursor.execute("UPDATE users SET role = 'admin' WHERE email = 'admin@example.com'")
    conn.commit()
    if cursor.rowcount > 0:
        print(f"Set admin@example.com to 'admin' role.")
except Exception as e:
    pass

conn.close()
print('Migration complete.')
